File: GOF/Prototype/java/src/maze/Maze.java
package maze;
import java.util.LinkedList;

public class Maze implements Cloneable {
    private LinkedList<Room> m_rooms = new LinkedList<>();

    public void addRoom(Room room)
    {
        m_rooms.addLast(room);
    }

    @Override
    public Object clone() throws CloneNotSupportedException {
        return super.clone();
    }
}

File: GOF/Prototype/cpp/maze/maze.cpp
//
// Created by gaojian on 19-2-27.
//

#include "maze.h"

using std::make_shared;

Maze::Maze(std::shared_ptr<Maze> maze)
{
	std::copy(maze->m_rooms.begin(), maze->m_rooms.end(), std::back_inserter(m_rooms));
}

void Maze::addRoom(std::shared_ptr<Room> room)
{
	m_rooms.push_back(room);
}

std::shared_ptr<Maze> Maze::clone()
{
	return shared_from_this();
}